compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -114,43 +152,36 @@ void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
         Operation *candidateSliceOp = allUsers[0];
         if (!(isa<tensor::InsertSliceOp, tensor::ParallelInsertSliceOp>(
                 candidateSliceOp))) {
-          producerOp = producerOp->getParentOfType<LoopLikeOpInterface>();
-          LLVM_DEBUG(
-              llvm::dbgs()
-              << "Going to reattempt fusion because didn't find "
-                 "tensor.insert_slice/tensor.parallel_insert_slice as the "
-                 "user of the compute op\n");
-          continue;
+          break;
         }
         std::optional<scf::SCFFuseConsumerOfSliceResult> fusedConsumer =
-            scf::tileAndFuseConsumerOfSlice(rewriter, candidateSliceOp);
+            scf::tileAndFuseConsumerOfSlice(rewriter, candidateSliceOp,
+                                            MutableArrayRef(loop));
         if (!fusedConsumer) {
-          producerOp = producerOp->getParentOfType<LoopLikeOpInterface>();
-          LLVM_DEBUG(llvm::dbgs()
-                     << "Failed to fuse any consumer op into the producer. "
-                        "Reattempt with loop-like parent operation.\n");
-          continue;
+          break;
         }
-        changed = true;
+        changedLocal = true;
+        changedGlobal = true;
         fusedConsumer->origConsumerOperand->getOwner()->erase();
-        Operation *fusedOp =
-            fusedConsumer->tiledAndFusedConsumerOperand->getOwner();
-        if (getAncestorInBlock(fusedOp, computeOp->getBlock()) != nullptr) {
-          // The consumer is fused all the way into the producer's block, so
-          // operate on this op from now on, but with reduced depth.
-          computeOp = fusedOp;
-          fuseDepth -= 1;
-        }
-        producerOp = fusedOp;
+        computeOp = fusedConsumer->tiledAndFusedConsumerOperand->getOwner();
         break;
-      } while (producerOp && producerOp->getParentOp() != funcOp);
+      } while (computeOp && computeOp->getParentOp() != funcOp);
+    }
+    if (changedLocal == false) break;
+    if (loopNestLength > 1) {
+      loop = cast<LoopLikeOpInterface>(
+          findPostOrderComputeOpInLoop<LoopLikeOpInterface>(
+              computeOp->getParentOfType<LoopLikeOpInterface>()));
     }
-    iter++;
+    loopNestLength--;
+    // Canonicalize before every iteration to enable more back-to-back fusion
+    // opportunities.
+    (void)applyPatternsGreedily(funcOp, canonicalizationPatterns);
   }
-  if (iter >= maxIterations) {
-    funcOp.emitOpError() << "Maximum number of iterations reached, consumer "
-                            "fusion is likely stuck in an infinite loop.";
-    return signalPassFailure();
+  if (changedGlobal) {
```

Comment:
Is this check correct? It looks to me like `changedGlobal` is set when fusions is performed sucessfully?

Reply:
Good catch! Updated.